File: richos/app/ui/tests/interruption.js
// **THE FAILURE CARD, WHEN THE FAILURE IS LOCAL** — the 2026-09-17 nightly's D2, rendered
// by the REAL renderer under WebKit, the engine Tauri ships on macOS.
//
// `docs/verification/2026-09-17-nightly-1.2.0-20260917.1-onscreen-audit.md` §D2. The
// published nightly held this in its own ledger:
//
//     "reason":"cognition protocol: \"Not logged in · Please run /login\""
//
// and showed him this:
//
//     Stopped after 1s
//     I hit a snag mid-thought and had to stop — say the word and I'll pick it back up.
//     Everything I'd already written above is saved.
//     [Pick it back up]
//
// Three untruths in four lines: a permanent condition called a passing hiccup, a promise
// about saved work when nothing had been written, and a retry control for a state no number
// of presses can clear. This suite makes each one unreachable on the surface the CEO sees.
//
// **THE BYTES ARE THE BACKEND'S, NOT THIS FILE'S.** Every sentence asserted below is scraped
// out of `crates/richos-core/src/interruption.rs` at run time by `lib/state-strings.js` —
// the same scrape `outage.js` and `affordances.js` use. Nothing here types a sentence, so a
// reworded `ceo_message()` moves this suite with it instead of leaving it green over copy
// the product no longer says.
//
// Run: node interruption.js   (or `npm test` for every suite in this directory)

"use strict";

const { loadPlaywright, openFixture, createRun, assert, assertEqual } = require("./lib/harness");
const { rustStrings, normalize } = require("./lib/state-strings");
const C = require("./lib/contrast");

const THREAD = "thr_interruption";
const TURN = "turn_interruption";

// ---------------------------------------------------------------------------------------
// THE SENTENCES, READ OUT OF THE RUST SOURCE
// ---------------------------------------------------------------------------------------

/// Find the one scraped `ceo_message` literal containing `needle`. Throws if there is not
/// exactly one — an ambiguous match means the source moved and this suite would otherwise
/// keep asserting against whichever it happened to pick.
function rustSentence(needle) {
  const hits = rustStrings()
    .map((s) => normalize(s.text))
    .filter((t) => t.includes(needle));
  const unique = Array.from(new Set(hits));
  assertEqual(
    unique.length,
    1,
    `expected exactly one Rust CEO sentence containing ${JSON.stringify(needle)}, found ${unique.length}`
  );
  return unique[0];
}

const NOT_SIGNED_IN = rustSentence("not connected to your Anthropic account");
const TRANSIENT = rustSentence("lost my connection to the part of me that thinks");
// The route is a WHOLE authored sentence, relayed verbatim — not a breadcrumb this file or
// the renderer wraps. `InterruptionCause::route`.
const ROUTE = rustSentence("under Account connection");

// The exact sentence `InterruptionRecord::new` writes when Rich had written nothing. It is
// a `format`-free literal in Rust, but it lives in `InterruptionRecord::new` rather than in
// a `ceo_message()` body, so the scraper does not see it and this is the one string the
// suite states. The core test `nothing_written_means_no_claim_about_a_saved_answer_is_made`
// pins the same bytes on the Rust side, so a change breaks there first.
const ONLY_HIS_WORDS = "Your message is saved. I hadn't written any of my answer yet.";

// The two generic sentences, which must NOT appear when a cause is known.
const GENERIC_BODY = "I hit a snag mid-thought and had to stop";
const GENERIC_NOTE = "Everything I'd already written above is saved.";

// ---------------------------------------------------------------------------------------
// FIXTURES
// ---------------------------------------------------------------------------------------

function snapshot(interruption, opts) {
  const withPartial = !!(opts && opts.withPartial);
  const b = (id, extra) =>
    Object.assign(
      {
        id,
        entityId: "northwind",
        threadId: THREAD,
        turnId: TURN,
        bindingRevision: 1,
        createdAt: 1789000000000,
        sequence: null,
        slot: "stream",
        visibility: "ceo",
      },
      extra || {}
    );
  const items = [
    Object.assign(b(TURN + ":user", { slot: "opening", createdAt: 1788000000000 }), {
      kind: "user_message",
      text: "What did I decide about nightly releases?",
      source: "text",
    }),
  ];
  // THE AUDIT'S CASE HAS NO PROSE ABOVE THE CARD, and that is load-bearing: the old note
  // claimed "everything I'd already written above is saved" with nothing above it at all.
  if (withPartial) {
    items.push(
      Object.assign(b(TURN + ":text:0", { sequence: 0 }), {
        kind: "rich_message",
        phase: "unknown",
        text: "Pulling the decision record now.",
      })
    );
  }
  items.push(
    Object.assign(b(TURN + ":duration", { slot: "terminal", createdAt: 1788000000000 }), {
      kind: "work_duration",
      state: "interrupted",
      startedAt: 1788000000000,
      endedAt: 1788000001000,
      activeMs: 1000,
    })
  );
  if (interruption) {
    items.push(
      Object.assign(b(TURN + ":interruption", { slot: "terminal", createdAt: 1788000001000 }), interruption)
    );
  }
  return { entityId: "northwind", threadId: THREAD, mode: "ceo", bindingRevision: 1, items };
}

/// The nightly's own turn, exactly as `richos-core` now projects it.
function notSignedInItem() {
  return {
    kind: "turn_interruption",
    cause: "not-signed-in",
    ceoMessage: NOT_SIGNED_IN,
    lossMessage: ONLY_HIS_WORDS,
    route: ROUTE,
    offersRetry: false,
  };
}

/// A lost pipe — the one class where asking again is honest advice.
function transientItem() {
  return {
    kind: "turn_interruption",
    cause: "transient",
    ceoMessage: TRANSIENT,
    lossMessage:
      "On disk: what you asked for is saved, the 31 characters of the answer that had " +
      "already arrived are saved. Not on disk: everything the session had worked out in " +
      "its head and had not yet said. Asking again starts that part over.",
    offersRetry: true,
  };
}

const cardText = (page) =>
  page.evaluate(() => {
    const card = document.querySelector(".tl-intervention");
    if (!card) return null;
    return {
      body: Array.from(card.querySelectorAll(".tl-intervention-body")).map((n) => n.textContent),
      notes: Array.from(card.querySelectorAll(".tl-intervention-note")).map((n) => n.textContent),
      action: card.querySelector(".tl-intervention-action")
        ? card.querySelector(".tl-intervention-action").textContent
        : null,
      actionCount: card.querySelectorAll(".tl-intervention-action").length,
    };
  });

const CARD_PARTS = ["tl-intervention-body", "tl-intervention-note", "tl-intervention-action"];

/// The whole fixture page, measured by `lib/contrast.js`'s shared probe — never arithmetic
/// of this file's own. Same reasoning as `outage.js`'s header records at length: a second
/// implementation that agrees today is one that will disagree silently later.
async function contrastOfPage(page, theme) {
  await page.evaluate(C.pageScript());
  const out = await page.evaluate((o) => window.__contrastProbe(o), {
    surface: "interruption-card",
    theme: theme,
  });
  return {
    measured: out.measuredPaths || [],
    failures: Object.values(out.failures),
    unresolvable: Object.values(out.unresolvable),
  };
}

/// Which of the card's own parts the probe measured, matched against the DOM rather than
/// against a guess at the probe's path format — so "the page is clean" can never mean "the
/// card was not on it".
async function cardNodes(page, measured) {
  const wanted = await page.evaluate((parts) => {
    const sel = parts.map((c) => "." + c).join(", ");
    return Array.from(document.querySelectorAll(sel)).map((el) => {
      const tag = el.tagName.toLowerCase();
      return el.id ? tag + "#" + el.id : tag + "." + String(el.className).split(/\s+/)[0];
    });
  }, CARD_PARTS);
  const missing = wanted.filter((w) => !measured.some((m) => m.indexOf(w) >= 0));
  return { wanted, missing };
}

// ---------------------------------------------------------------------------------------

async function main() {
  const { webkit } = loadPlaywright();
  const browser = await webkit.launch();
  const run = createRun("nightly D2 — a failed turn says the cause the app is holding");

  const page = await openFixture(browser);

  // =====================================================================================
  // THE NIGHTLY'S CASE
  // =====================================================================================

  await run.check("the sign-in case replaces the generic card with the backend's sentence", async () => {
    await page.evaluate((s) => window.__render(s), snapshot(notSignedInItem()));
    await page.waitForSelector(".tl-intervention");
    const c = await cardText(page);
    assert(c, "no failure card rendered at all");
    assertEqual(c.body, [NOT_SIGNED_IN], "the body is the authored sentence, verbatim");
    assert(
      !c.body.concat(c.notes).some((t) => t.includes(GENERIC_BODY)),
      "UNTRUTH 1 — the generic snag sentence must be replaced, not appended to"
    );
    return "body = the authored sign-in sentence";
  });

  await run.check("it names the account and the route that exists", async () => {
    const c = await cardText(page);
    const whole = c.body.concat(c.notes).join(" ");
    assert(whole.includes("Anthropic account"), "the account is never mentioned: " + whole);
    assert(
      c.notes.indexOf(ROUTE) >= 0,
      "the authored route sentence must be relayed verbatim: " + JSON.stringify(c.notes)
    );
    return "names the account + Settings → Account connection";
  });

  await run.check("UNTRUTH 2 — it claims no saved answer when none was written", async () => {
    const c = await cardText(page);
    assert(
      !c.notes.some((n) => n.includes(GENERIC_NOTE)),
      '"Everything I\'d already written above is saved" must be gone: ' + JSON.stringify(c.notes)
    );
    assert(
      !c.notes.some((n) => n.includes("already written above")),
      "no claim about prose above the card, because there is none: " + JSON.stringify(c.notes)
    );
    assert(
      c.notes.some((n) => n.includes(ONLY_HIS_WORDS)),
      "the one true thing — HIS message is saved — must still be said: " + JSON.stringify(c.notes)
    );
    // And the screen agrees with the claim: there is genuinely no Rich prose above it.
    const prose = await page.evaluate(() => document.querySelectorAll(".tl-rich, .tl-message--rich").length);
    assertEqual(prose, 0, "the fixture must have no answer above the card for this to mean anything");
    return "no saved-answer claim; his message named; 0 prose rows above";
  });

  await run.check("UNTRUTH 3 — there is no retry control on a state a retry cannot clear", async () => {
    const c = await cardText(page);
    assertEqual(c.actionCount, 0, "a button that cannot work is worse than no button");
    assertEqual(c.action, null, "no control at all, not a disabled one with the same promise");
    return "0 retry controls";
  });

  // =====================================================================================
  // THE POSITIVE CONTROL — without it, "no button" could be "the card never renders one"
  // =====================================================================================

  await run.check("a transient failure DOES get the retry control, and the one verb", async () => {
    await page.evaluate((s) => window.__render(s), snapshot(transientItem(), { withPartial: true }));
    await page.waitForSelector(".tl-intervention");
    const c = await cardText(page);
    assertEqual(c.body, [TRANSIENT], "the body is the authored transient sentence");
    assertEqual(c.actionCount, 1, "exactly one control");
    assertEqual(c.action, "Pick it back up", "same verb as every other failure — one thing to learn");
    assert(
      c.notes.some((n) => n.includes("31 characters of the answer")),
      "what IS on disk is counted: " + JSON.stringify(c.notes)
    );
    return "Pick it back up — present, and the partial reply counted";
  });

  await run.check("an unclassified turn still falls back to the generic card", async () => {
    // A turn written before 2026-09-17 carries no cause. It must render exactly as it
    // always did rather than losing its card — the record is absent, not the failure.
    await page.evaluate((s) => window.__render(s), snapshot(null));
    await page.waitForSelector(".tl-intervention");
    const c = await cardText(page);
    assert(
      c.body.some((t) => t.includes(GENERIC_BODY)),
      "the legacy card must survive for records that carry no cause: " + JSON.stringify(c.body)
    );
    assertEqual(c.actionCount, 1, "and it keeps its control");
    return "legacy record -> generic card, unchanged";
  });

  // =====================================================================================
  // CONTRAST — WCAG AA in BOTH themes, computed, never eyeballed
  // =====================================================================================

  for (const theme of ["dark", "light"]) {
    await run.check(`the card meets WCAG AA in the ${theme} theme`, async () => {
      await page.evaluate((t) => {
        document.documentElement.setAttribute("data-theme", t);
      }, theme);
      await page.evaluate((s) => window.__render(s), snapshot(notSignedInItem()));
      await page.waitForSelector(".tl-intervention");
      const r = await contrastOfPage(page, theme);
      const nodes = await cardNodes(page, r.measured);
      assertEqual(
        nodes.missing.length,
        0,
        `the card's own parts were not measured: ${JSON.stringify(nodes.missing)}`
      );
      assertEqual(
        r.unresolvable.length,
        0,
        `a color the probe could not resolve is a failure to prove: ${JSON.stringify(r.unresolvable)}`
      );
      assertEqual(
        r.failures.length,
        0,
        `WCAG AA failures in ${theme}: ${JSON.stringify(r.failures)}`
      );
      return `${theme}: ${r.measured.length} node(s) measured, ${nodes.wanted.length} card part(s), 0 failures`;
    });
  }

  // =====================================================================================
  // THE TYPE FLOOR — §15: nothing a person is expected to read below 14px
  // =====================================================================================

  await run.check("no part of the card sits below the 14px floor", async () => {
    const sizes = await page.evaluate((parts) => {
      const sel = parts.map((c) => "." + c).join(", ");
      return Array.from(document.querySelectorAll(sel)).map((el) => ({
        cls: String(el.className),
        px: parseFloat(getComputedStyle(el).fontSize),
      }));
    }, CARD_PARTS);
    assert(sizes.length > 0, "nothing measured");
    const small = sizes.filter((s) => s.px < 14);
    assertEqual(small.length, 0, `below the floor: ${JSON.stringify(small)}`);
    return sizes.map((s) => `${s.cls.split(/\s+/)[0]}=${s.px}px`).join(" ");
  });

  await page.evaluate(() => document.documentElement.setAttribute("data-theme", "dark"));
  await browser.close();
  // `report()` returns the FAILED COUNT, not a verdict. Same form as every other suite here.
  process.exit(run.report() > 0 ? 1 : 0);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: richos/app/crates/richos-core/src/interruption.rs
//! **WHY A TURN ENDED WITHOUT FINISHING — the vocabulary, and the one place it is decided.**
//!
//! `docs/verification/2026-09-17-nightly-1.2.0-20260917.1-onscreen-audit.md` §D2. The
//! published nightly wrote this to its own ledger:
//!
//! ```text
//! {"event":"TurnInterrupted","turn_id":"turn_afe3ad4a…",
//!  "reason":"cognition protocol: \"Not logged in · Please run /login\"", …}
//! ```
//!
//! and showed the CEO this:
//!
//! > **Stopped after 1s**
//! > I hit a snag mid-thought and had to stop — say the word and I'll pick it back up.
//! > Everything I'd already written above is saved.
//! > \[Pick it back up\]
//!
//! **The app had the exact, actionable cause and showed him something else.** Three separate
//! untruths in four lines: a permanent condition described as a passing hiccup; a promise
//! about saved work when nothing had been written; and a retry control for a state no number
//! of retries can clear. The audit's own note on why this outlives its harness is the reason
//! this module exists: *"not logged in is not an exotic state — it is where every expiring
//! subscription session lands."*
//!
//! # WHAT THIS MODULE DECIDES, AND WHAT IT REFUSES TO DECIDE
//!
//! It answers one question — *"which KIND of ending was this?"* — and it answers
//! [`InterruptionCause::Unknown`] rather than guessing. It is the sibling of
//! [`crate::upstream`], which answers the same shape of question about the model API, and it
//! is deliberately NOT an arm of that enum: a `claude` that refuses locally because nobody is
//! signed in never reached Anthropic, so filing it under "upstream" would be the exact
//! dishonesty that module's header warns against.
//!
//! # IT IS KEYED ON THE VENDOR'S OWN CONSTANTS, READ OUT OF THE SHIPPED BUNDLE
//!
//! There is no status code or error tag on this path — `native.rs` builds
//! `CognitionError::Protocol` from the `result` frame's `result` string, so the only thing
//! RichOS holds is the sentence Claude Code wrote. That sentence is not guessed at here. It
//! was read verbatim out of the installed bundle
//! (`~/.local/share/claude/versions/2.1.274`, 2026-09-17), where the whole family sits
//! together as adjacent constants:
//!
//! ```js
//! gVe="Not logged in \xB7 Please run /login",
//! HEe="Invalid API key \xB7 Fix external API key",
//! Qwn="Invalid auth token \xB7 Fix external auth token",
//! Zwn="Invalid ANTHROPIC_CUSTOM_HEADERS \xB7 Fix the environment variable",
//! Jwn="Invalid request header from the environment \xB7 Fix the environment variable"
//! ```
//!
//! **Those five are two different problems with two different actors**, which is why they are
//! two arms below. The first is the CEO's own account. The other four are a credential
//! somebody put in this machine's environment, and the person who can fix one of those is
//! whoever set RichOS up — telling the CEO to reconnect his account would send him to a
//! screen that cannot help.
//!
//! **A reworded vendor string falls to [`InterruptionCause::Unknown`], and that is the safe
//! direction**: Unknown says plainly that RichOS does not know why, which is true, and offers
//! no route that might be wrong. It never silently becomes the arm that looked closest.
//!
//! # NOTHING HERE EVER CLAIMS SOMETHING SURVIVED THAT DID NOT
//!
//! [`InterruptionRecord::new`] builds its loss sentence from [`crate::upstream::TurnLoss`] —
//! counts read off the ledger at the instant of the failure — and when nothing of Rich's had
//! been written it says nothing about saved work at all. The sentence the audit caught,
//! *"Everything I'd already written above is saved"*, is unreachable from here by
//! construction.

use serde::{Deserialize, Serialize};

use crate::upstream::TurnLoss;

/// The CEO's own sign-in, missing or expired. Read verbatim from the shipped bundle; see
/// the module header.
const VENDOR_NOT_SIGNED_IN: &str = "Not logged in";

/// A credential that is PRESENT and rejected — an API key, an auth token, or a header set in
/// this machine's environment. Four vendor constants, all of them somebody else's to fix.
const VENDOR_CREDENTIAL_REJECTED: [&str; 4] = [
    "Invalid API key",
    "Invalid auth token",
    "Invalid ANTHROPIC_CUSTOM_HEADERS",
    "Invalid request header from the environment",
];

/// RichOS's own sentence when the CEO's stop reached the child (`native.rs`, the `Cancel`
/// arm). Ours, not the vendor's, so matching it is matching our own structure.
const OUR_STOPPED_AT_YOUR_REQUEST: &str = "stopped at your request";

/// `CognitionError::Io`'s `Display` prefix. A broken pipe, a closed child, a read that
/// failed — the class where asking again is a real plan.
const OUR_IO_PREFIX: &str = "cognition io:";

/// Which kind of ending this was.
///
/// Five arms, and the fifth is the honest one. Compare [`crate::upstream::UpstreamFault`],
/// which is shaped the same way and for the same reason.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum InterruptionCause {
    /// Nobody is signed in to the Anthropic account RichOS thinks with, or the session
    /// expired. **Permanent until somebody signs in**; retrying cannot clear it.
    NotSignedIn,
    /// A credential IS configured on this machine and was rejected. Also permanent until
    /// changed, and the person who can change it is whoever set RichOS up — not the CEO.
    CredentialRejected,
    /// The CEO stopped it. Not a failure at all, and must never be dressed as one.
    StoppedByCeo,
    /// A pipe, a child that went away, a read that failed. **The one arm where asking
    /// again is a plan**, which is exactly why it is not the default.
    Transient,
    /// Something RichOS has no story for. Named rather than guessed.
    Unknown,
}

impl InterruptionCause {
    /// The machine-readable tag stored in the ledger and read back by the projection. A tag
    /// this build does not recognize reads back as [`InterruptionCause::Unknown`], which is
    /// the safe direction for a record written by a newer RichOS.
    pub fn tag(self) -> &'static str {
        match self {
            InterruptionCause::NotSignedIn => "not-signed-in",
            InterruptionCause::CredentialRejected => "credential-rejected",
            InterruptionCause::StoppedByCeo => "stopped-by-ceo",
            InterruptionCause::Transient => "transient",
            InterruptionCause::Unknown => "unknown",
        }
    }

    pub fn from_tag(tag: &str) -> InterruptionCause {
        match tag {
            "not-signed-in" => InterruptionCause::NotSignedIn,
            "credential-rejected" => InterruptionCause::CredentialRejected,
            "stopped-by-ceo" => InterruptionCause::StoppedByCeo,
            "transient" => InterruptionCause::Transient,
            _ => InterruptionCause::Unknown,
        }
    }

    /// **THE CEO-FACING LINE, and there is deliberately no shared arm.**
    ///
    /// Named `ceo_message` for the reason [`crate::upstream::UpstreamFault::ceo_message`]
    /// gives: `app/ui/tests/lib/state-strings.js` scrapes the product's CEO-facing sentences
    /// out of source, and under `app/crates` the only shape it can see is a literal inside a
    /// function with this name.
    ///
    /// **Every one of them reads the same spoken as written.** RichOS is voice-first, so a
    /// sentence that needs a screen to make sense is a sentence that fails half the time it
    /// is delivered. No paths, no error codes, no punctuation doing structural work.
    ///
    /// **None of them promises a retry.** Whether asking again can help is
    /// [`InterruptionCause::offers_retry`], and the sentence and the control are driven by
    /// the same fact so they cannot disagree — which is precisely how the nightly ended up
    /// inviting the CEO to press a button that could not work.
    pub fn ceo_message(self) -> &'static str {
        match self {
            // THE ROUTE IS IN THE SENTENCE, not beside it. It was a second string returned
            // by a `route()` method until `affordances.js` refused the row: under
            // `app/crates` the state registry can only see literals inside a function
            // called `ceo_message`, so a route authored anywhere else would have reached his
            // screen while being invisible to the one inventory that asks whether he can act
            // on what he is told. That is the hole this module's own header objects to, and
            // the fix is to have one authored sentence rather than a way around the scrape.
            //
            // It also reads better aloud. `Settings → Account connection` was the first
            // form, and an arrow is read aloud as nothing at all.
            InterruptionCause::NotSignedIn => {
                "I couldn't start that, because I'm not connected to your Anthropic account \
                 right now — either nobody has signed in on this Mac yet, or the sign-in ran \
                 out. You can connect it in Settings, under Account connection, and I'll \
                 pick this straight back up."
            }
            InterruptionCause::CredentialRejected => {
                "I couldn't start that. There is an account credential set up on this Mac and \
                 Anthropic turned it down, so the request never left the machine. This one \
                 needs whoever set RichOS up — it isn't something you can fix from here, and \
                 asking me again won't change it."
            }
            InterruptionCause::StoppedByCeo => {
                "Stopped, as you asked. Nothing is running, and nothing of yours was lost."
            }
            InterruptionCause::Transient => {
                "I lost my connection to the part of me that thinks, partway through. That \
                 kind of thing usually clears on its own, so asking again is worth a try."
            }
            InterruptionCause::Unknown => {
                "That stopped before I finished, and I can't tell you why — I don't recognize \
                 what came back. Asking again is reasonable; if it stops the same way, it \
                 needs whoever set RichOS up."
            }
        }
    }

    /// **Is asking again a plan, or a false promise?**
    ///
    /// The whole of D2's harm sits in this method. `false` means the surface must not draw a
    /// retry control: a button that cannot work is worse than no button, because pressing it
    /// and watching nothing happen is how somebody stops believing what the app tells them.
    pub fn offers_retry(self) -> bool {
        match self {
            InterruptionCause::NotSignedIn | InterruptionCause::CredentialRejected => false,
            // Nothing failed. There is nothing to retry, and his words go back in the box
            // by the ordinary route rather than through a failure card's control.
            InterruptionCause::StoppedByCeo => false,
            InterruptionCause::Transient | InterruptionCause::Unknown => true,
        }
    }

}

/// Decide the cause from the reason string the ledger stores.
///
/// **The input is deliberately the SAME string `Ledger::interrupt_turn` is given**, so the
/// record and the reason can never describe two different events. The caller classifies at
/// the failure boundary and writes both.
///
/// Order matters in one place only: the vendor ships
/// `"API Error: 401 Invalid API key · Please run /login"`, which contains BOTH families'
/// wording. It is a rejected credential, not a missing sign-in, so the sign-in test is the
/// narrower string and the credential test runs where it can still win.
pub fn classify(reason: &str) -> InterruptionCause {
    if reason.contains(OUR_STOPPED_AT_YOUR_REQUEST) {
        return InterruptionCause::StoppedByCeo;
    }
    if VENDOR_CREDENTIAL_REJECTED.iter().any(|needle| reason.contains(needle)) {
        return InterruptionCause::CredentialRejected;
    }
    if reason.contains(VENDOR_NOT_SIGNED_IN) {
        return InterruptionCause::NotSignedIn;
    }
    if reason.contains(OUR_IO_PREFIX) {
        return InterruptionCause::Transient;
    }
    InterruptionCause::Unknown
}

/// **What is written to the ledger when a turn ends without finishing — and it carries the
/// SENTENCES, not just the cause.**
///
/// Same reasoning as [`crate::upstream::UpstreamRecord`], which this mirrors field for
/// field where the fields mean the same thing: the record is the evidence of what the CEO
/// was told, at the moment he was told it. A later build that rewords
/// [`InterruptionCause::ceo_message`] must not be able to rewrite what a six-month-old
/// thread shows him.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InterruptionRecord {
    /// [`InterruptionCause::tag`].
    pub cause: String,
    /// [`InterruptionCause::ceo_message`] as it stood when this happened.
    pub ceo_message: String,
    /// What was on disk and what was not — **absent entirely when nothing of Rich's had
    /// been written**, which is the sentence the nightly got wrong.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub loss_message: Option<String>,
    /// [`InterruptionCause::offers_retry`], stored rather than re-derived so the control the
    /// CEO sees on a reload is the control he saw at the time.
    pub offers_retry: bool,
}

impl InterruptionRecord {
    /// Assemble the record from the two things that are true at the failure boundary.
    ///
    /// # THE LOSS SENTENCE, AND WHEN THERE ISN'T ONE
    ///
    /// * **Rich had written something** — a partial reply, or a recorded action — so the
    ///   full statement from [`TurnLoss::ceo_message`] applies: what is on disk, what is
    ///   not, and what that means for asking again.
    /// * **Rich had written nothing, but the CEO's message is durable** — one short true
    ///   sentence about HIS words, which is the thing he actually wants to know. It says
    ///   nothing about an answer, because there was no answer.
    /// * **Nothing at all is on disk** — no sentence. Silence is the honest output; a
    ///   reassurance about nothing is what produced D2.
    pub fn new(cause: InterruptionCause, loss: &TurnLoss) -> InterruptionRecord {
        let rich_wrote_something = loss.partial_reply_chars > 0 || loss.actions_recorded > 0;
        let loss_message = if rich_wrote_something {
            Some(loss.ceo_message())
        } else if loss.prompt_is_durable {
            Some("Your message is saved. I hadn't written any of my answer yet.".to_string())
        } else {
            None
        };
        InterruptionRecord {
            cause: cause.tag().to_string(),
            ceo_message: cause.ceo_message().to_string(),
            loss_message,
            offers_retry: cause.offers_retry(),
        }
    }
}
